src/storageddbus.c

```diff
@@ -0,0 +1,360 @@
+  GSocket *socket = NULL;
+  GIOStream *io = NULL;
+  GError *error = NULL;
+  int fds[2] = { -1, -1 };
+  gboolean ret = FALSE;
+  sd_bus *bus;
+  int r;
+
+  if (socketpair (PF_UNIX, SOCK_STREAM, 0, fds) < 0)
+    {
+      g_critical ("Couldn't create socket pair: %s", g_strerror (errno));
+      goto out;
+    }
+
+  r = sd_bus_get_bus_id (system, &server_id);
+  if (r < 0)
+    {
+      g_critical ("Couldn't connect to system bus: %s", g_strerror (-r));
+      goto out;
+    }
+
+  /* Open an sd_bus for this FD */
+  if (sd_bus_new (&bus) < 0)
+    g_return_val_if_reached (FALSE);
+  if (sd_bus_set_fd (bus, fds[0], fds[0]) < 0)
+    g_return_val_if_reached (FALSE);
+  fds[0] = -1; /* Claimed */
+  if (sd_bus_set_server (bus, 1, server_id) < 0)
+    g_return_val_if_reached (FALSE);
+  if (sd_bus_set_anonymous (bus, 1) < 0)
+    g_return_val_if_reached (FALSE);
+  if (sd_bus_attach_event (bus, NULL, SD_EVENT_PRIORITY_NORMAL) < 0)
+    g_return_val_if_reached (FALSE);
+  r = sd_bus_start (bus);
+  if (r < 0)
+    {
+      g_critical ("Couldn't start proxy bus: %s", g_strerror (r));
+      goto out;
+    }
+
+  if (sd_bus_add_filter (system, &filter_in, on_message_filter, bus) < 0)
+    g_return_val_if_reached (FALSE);
+  if (sd_bus_add_filter (bus, &filter_out, on_message_filter, system) < 0)
+    g_return_val_if_reached (FALSE);
+
+  socket = g_socket_new_from_fd (fds[1], &error);
+  if (!socket)
+    {
+      g_critical ("Couldn't create socket from fd: %s", error->message);
+      g_clear_error (&error);
+      goto out;
+    }
+
+  fds[1] = -1; /* Claimed */
+
+  io = G_IO_STREAM (g_socket_connection_factory_create_connection (socket));
+  flags = G_DBUS_CONNECTION_FLAGS_AUTHENTICATION_CLIENT | G_DBUS_CONNECTION_FLAGS_MESSAGE_BUS_CONNECTION;
+  g_dbus_connection_new (io, NULL, flags, NULL, NULL, on_ready_get_result, &result);
+
+  /*
+   * We are actually a sync function, but we spin the main loop to allow
+   * the GDBus connection and sd_bus proxy to start communicating properly.
+   */
+
+  while (result == NULL)
+    g_main_context_iteration (NULL, TRUE);
+
+  *connection = g_dbus_connection_new_finish (result, &error);
+  if (*connection)
+    {
+      ret = TRUE;
+      *proxy = bus;
+      bus = NULL;
+    }
+  else
+    {
+      g_critical ("Couldn't create proxy GDBus to sd-bus connection: %s", error->message);
+      g_clear_error (&error);
+    }
+
+out:
+  if (fds[0] >= 0)
+    close (fds[0]);
+  if (fds[1] >= 0)
+    close (fds[1]);
+  g_clear_object (&result);
+  g_clear_object (&socket);
+  g_clear_object (&io);
+
+  if (bus)
+    sd_bus_unref (bus);
+
+  return ret;
+}
+
+static gboolean
+open_system_bus (sd_bus **system)
+{
+  int r;
+
+  g_assert (system != NULL);
+
+  r = sd_bus_default_system (system);
+  if (r < 0)
+    {
+      g_critical ("Couldn't open system DBus bus: %s", g_strerror (-r));
+      return FALSE;
+    }
+
+  if (sd_bus_attach_event (*system, NULL, SD_EVENT_PRIORITY_NORMAL) < 0)
```

Comment:
I cannot find any piece of documentation for this function, but shouldn't this pass some event to it?